src/wp-includes/blocks.php

```diff
@@ -797,34 +802,48 @@ function make_before_block_visitor( $context ) {
-				if ( 'first_child' === $relative_position ) {
-					$hooked_block_markup = get_comment_delimited_block_content( $hooked_block_type, array(), '' );
-					/** This filter is documented in wp-includes/blocks.php */
-					$markup .= apply_filters( 'inject_hooked_block_markup', $hooked_block_markup, $hooked_block_type, $relative_position, $parent, $context );
-				}
+			$relative_position  = 'first_child';
+			$anchor_block_type  = $parent['blockName'];
+			$hooked_block_types = array_keys( get_hooked_blocks( $anchor_block_type, $relative_position ) );
+			/**
+			 * Filters the list of hooked block types for a given anchor block type and relative position.
+			 *
+			 * @since 6.4.0
+			 *
+			 * @param string[]                $hooked_block_types  The list of hooked block types.
+			 * @param string                  $relative_position   The relative position of the hooked blocks.
+			 *                                                     Can be one of 'before', 'after', 'first_child', or 'last_child'.
+			 * @param string                  $anchor_block_type   The anchor block type.
+			 * @param WP_Block_Template|array $context             The block template, template part, or pattern that the anchor block belongs to.
+			 */
+			$hooked_block_types = apply_filters( 'hooked_block_types', $hooked_block_types, $relative_position, $anchor_block_type, $context );
+			foreach ( $hooked_block_types as $hooked_block_type ) {
+				$hooked_block_markup = get_comment_delimited_block_content( $hooked_block_type, array(), '' );
+				/** This filter is documented in wp-includes/blocks.php */
+				$markup .= apply_filters( 'inject_hooked_block_markup', $hooked_block_markup, $hooked_block_type, $relative_position, $parent, $context );
 			}
 		}
 
-		$hooked_blocks = get_hooked_blocks( $block['blockName'] );
-		foreach ( $hooked_blocks as $hooked_block_type => $relative_position ) {
-			if ( 'before' === $relative_position ) {
-				$hooked_block_markup = get_comment_delimited_block_content( $hooked_block_type, array(), '' );
-				/**
-				 * Filters the serialized markup of a hooked block.
-				 *
-				 * @since 6.4.0
-				 *
-				 * @param string                  $hooked_block_markup The serialized markup of the hooked block.
-				 * @param string                  $hooked_block_type   The type of the hooked block.
-				 * @param string                  $relative_position   The relative position of the hooked block.
-				 *                                                     Can be one of 'before', 'after', 'first_child', or 'last_child'.
-				 * @param array                   $block               The anchor block.
-				 * @param WP_Block_Template|array $context             The block template, template part, or pattern that the anchor block belongs to.
-				 */
-				$markup .= apply_filters( 'inject_hooked_block_markup', $hooked_block_markup, $hooked_block_type, $relative_position, $block, $context );
-			}
+		$relative_position  = 'before';
```

Comment:
Minor, but I wonder if the various `$relative_position` values should be constants (something like `WP_Block_Hooks::RELATIVE_POSITION_BEFORE`, `WP_Block_Hooks::RELATIVE_POSITION_FIRST_CHILD` ...etc). This could reduce some variable declaration and make it more straightforward/clear for extension use.

Reply:
Ah yeah, good point. I think it hadn't occurred to me since we need to use the position values verbatim in `block.json` -- but it might make sense to add constants in the code 👍

Reply:
did we update the `block.json` schema so that these values can be drawn from an auto-completed list with inline documentation?

Reply:
This is what we have in the schema:

https://github.com/WordPress/gutenberg/blob/aa938166f28bcfff6c7b8d1caac5a95f78852f97/schemas/json/block.json#L686-L695

Reply:
Speaking of schema, I opened https://github.com/WordPress/wordpress-develop/pull/5278 to align the schemas between Gutenberg and WordPress core.